feat(toolbar layout): implements toolbar layout

mattnolting opened the pull request:
closes #136 

Working PR
Based on past experience, relying only on flexbox for responsive layout produces suboptimal results. Therefore, I included a few responsive modifiers to better control layout. 

Questions:
- Should the title of this component be genericized? In talking with @jgiardino about accessibility compliance, toolbars are limited to buttons or controls https://www.w3.org/TR/wai-aria-1.1/#toolbar. It seems this layout would benefit from a more generic name.

patternfly-build commented:
Deploy preview for *pf-next* ready!

Built with commit 2df04386936b70b11ad59a220b875ba81f5efd63

https://deploy-preview-391--pf-next.netlify.com

michael-coker commented on src/patternfly/layouts/Toolbar/docs/code.md:
```diff
@@ -0,0 +1,29 @@
+| `.pf-l-toolbar__section` | `<section>` |  Initiates a toolbar section. Use this when vertically stacking toolbar sections. |
+| `.pf-l-toolbar__group` | `<div>` |  Initiates a toolbar group. **Required.** |
+| `.pf-l-toolbar__item` | `<div>` |  Initiates a toolbar item. **Required.** |
+| `.pf-m-wrap` | `.pf-l-toolbar` |  Sets toolbar items to wrap. |
```
Should this read `Sets toolbar groups to wrap.`? Since the class is applies to `.pf-l-toolbar` and it looks like the required direct descendent of that is `.pf-l-toolbar__group`.

michael-coker commented on src/patternfly/layouts/Toolbar/docs/code.md:
```diff
@@ -0,0 +1,29 @@
+| `.pf-l-toolbar__group` | `<div>` |  Initiates a toolbar group. **Required.** |
+| `.pf-l-toolbar__item` | `<div>` |  Initiates a toolbar item. **Required.** |
+| `.pf-m-wrap` | `.pf-l-toolbar` |  Sets toolbar items to wrap. |
+| `.pf-m-nowrap` | `.pf-l-group` |  Sets toolbar group items to nowrap. |
```
Should the class it applies to be `.pf-l-toolbar__group`?

michael-coker commented on src/patternfly/layouts/Toolbar/styles.scss:
```diff
@@ -0,0 +1,173 @@
+  /* stylelint-enable */
+
+  // Toolbar base styling
+  flex-flow: row wrap;
```
If the toolbar isn't supposed to wrap by default, and you apply `.pf-m-wrap` for it to wrap, should this be `nowrap` instead (or just remove the entire line since `row` and `nowrap` are defaults)

michael-coker replied:
You could also just use `flex-wrap: wrap` here.

michael-coker commented on src/patternfly/layouts/Toolbar/styles.scss:
```diff
@@ -0,0 +1,173 @@
+
+  // Set flex items to align baseline
+  .pf-m-align-items-baseline {
+    align-items: first baseline;
```
is `first` a typo?

mattnolting replied:
It's not a typo, but don't think it's necessary.
https://developer.mozilla.org/en-US/docs/Web/CSS/align-items
https://drafts.csswg.org/css-flexbox-1/#flex-baselines

michael-coker replied:
cool! I haven't seen that before. thanks for the reference.

michael-coker commented on src/patternfly/layouts/Toolbar/styles.scss:
```diff
@@ -0,0 +1,173 @@
+  }
+
+  // Set flex items to align top
+  .pf-m-align-items-top {
```
Have we considered using `start` in the class instead of `top`? (Also using `end` instead of `bottom`). For the toolbar, I don't think it matters, but we're using the css property `align-items` in the class. What about a future component or layout where you can change the `flex-direction`, and also have a `.pf-m-align-items-*` modifier? In that case, `flex-start` won't always mean "top" and `flex-end` won't always mean "bottom". Seems like we could keep this modifier usage more consistent and meaningful by using the actual value for the property instead of the visual placement.

mattnolting replied:
Agree

michael-coker commented on src/patternfly/layouts/Toolbar/styles.scss:
```diff
@@ -0,0 +1,173 @@
+  }
+
+  // Set flex item to occupy available space
+  .pf-m-stretch {
```
Have we considered using `grow` here instead of `stretch`? I think it's just as meaningful and less confusing because `stretch` is a value for `align-items`. That would allow us to use a `.pf-m-stretch` class throughout patternfly to represent `align-items: stretch`.

mattnolting replied:
Agree

michael-coker commented on src/patternfly/layouts/Toolbar/styles.scss:
```diff
@@ -0,0 +1,173 @@
+  }
+
+  // Responsive classes
+  .pf-m-stretch-on {
```
just a formatting thing, but I would nest this under `.pf-m-stretch`. Same for `.pf-m-align-right-on` below.

michael-coker commented on src/patternfly/layouts/Toolbar/examples/toolbar-section-example.hbs:
```diff
@@ -0,0 +1,42 @@
+        {{#> badge pf-c-badge--modifier="pf-m-unread"}}3{{/badge}}
+      {{/toolbar-item}}
+      {{#> toolbar-item}}
+        {{#> badge pf-c-badge--modifier="pf-m-unread"}}54321{{/badge}}
```
Per the badge spec (can find it in https://github.com/patternfly/patternfly-next/issues/22), the count truncates after 3 digits with a `+`, so this should read `999+`. @mceledonia please correct me if I'm wrong :)

michael-coker commented on src/patternfly/layouts/Toolbar/examples/toolbar-section-example.hbs:
```diff
@@ -0,0 +1,42 @@
+    {{/toolbar-group}}
+    {{#> toolbar-group newcontent  pf-l-toolbar--modifier="pf-m-align-right-on-md"}}
+      {{#> toolbar-item newcontent}}
+        <a href="#">Go to badges</a>
```
Should this link to `/components/badge/examples/`?

michael-coker commented on src/patternfly/layouts/Toolbar/styles.scss:
```diff
@@ -0,0 +1,173 @@
+
+    // Add margin-top for stacked sections
+    &:not(:first-child) {
+      margin-top: var(--pf-l-toolbar__section--MarginTop);
```
Nitpicky, but you're applying a `margin` before each section element here, and excluding the `:first-child`, but on groups and group items, you're applying the `margin` after each element, and excluding the `:last-child`.

Is there a reason behind applying the `margin` before sections? If not, I would expect the margin to come after, and you exclude `:last-child` like groups and group items.

mattnolting replied:
As a rule, I try to always use :first-child as the first option. In this case, using `:first-child` with `left-margin` on group items causes misalignment when they wrap. I could change it, what do you think @andresgalante?

andresgalante replied:
@michael-coker I had the same reaction. Since section are optional to create new rows then I think it's a good move to have a margin-top on any row that's not the first one.

michael-coker commented on src/patternfly/layouts/Toolbar/examples/toolbar-section-example.hbs:
```diff
@@ -0,0 +1,42 @@
+        {{#> badge pf-c-badge--modifier="pf-m-unread"}}3{{/badge}}
+      {{/toolbar-item}}
+      {{#> toolbar-item}}
+        {{#> badge pf-c-badge--modifier="pf-m-unread"}}543+{{/badge}}
```
Per the spec @mceledonia gave, I think this might be technically correct? Though logically, I would expect that as the value in a badge hits 999, it is represented as `999+` until the value drops below that number again.

andresgalante commented:
Hi @mattnolting Sorry, I think I didn't scoped it enough in the issue (again).

The layout should be empty. if you want to create examples those should be demos.

The tool bar groups should not have a thing on the right. Designers specially asked that toolbars group on the lest.

If a user wishes to have a thing on the right and a thing on the left they can use a split with a tool bar inside.

mattnolting commented:
@andresgalante Got my demos and layouts mixed up! I've updated the layout to include only simple items. I left `pf-l-toolbar__section` assuming there will be a need in the future when more complex toolbar layouts are designed, will remove if you prefer. The idea is that multiple `sections` can be used in favor of multiple `toolbars`. Also an assumption, but it seems very reasonable that there will be a need in the near future.

mattnolting commented:
Based on the structure, this looks more like `pf-l-flex` layout, which could easily be applied to a toolbar demo, footer actions, etc, as it employs the same concepts as https://pf-next.com/layouts/Level/examples/, https://pf-next.com/layouts/Split/examples/, https://pf-next.com/layouts/Stack/examples/. However, that may be another, separate issue..

andresgalante commented:
@mattnolting I understand your concern, an abstraction to a `pf-flex` layout would cope with these usecases, a grid would probably also do the same. My intention was to build the layouts based on what they are and not how they are build. 

That way we don't attach the API to the implementation, and if down the road there is a better CSS way to build a split, a toolbar or a gallery then we can change the implementation and keep the same classname API.

The cost of course is repetition of code.

@mattnolting Does it make sense? What do you think?

@michael-coker @srambach @jgiardino @matthewcarleton @mindreeper2420 What do you think?

andresgalante commented on src/patternfly/layouts/Toolbar/styles.scss:
```diff
@@ -0,0 +1,146 @@
+  // Item
+  &__item {
+    // Setting line-height in order to vertically align properly. Ex: badge
+    line-height: 1;
```
I wouldn't change the line height

andresgalante commented on src/patternfly/layouts/Toolbar/styles.scss:
```diff
@@ -0,0 +1,146 @@
+    flex-wrap: nowrap;
+  }
+
+  // Set flex items to align top
```
Do we need these modifiers? I think that by default the tool bar items should align to the center.

If we do start adding this kind of modifiers maybe is a good case to do the `pf-flex` layout.

mattnolting replied:
Agreed

mattnolting commented:
@andresgalante I think there's value in use case layouts, a grid layout system and a flexbox layout system. I see grid as mostly a page layout system (although it can also be used in a toolbars, cards, etc for example) and flex as a fluid approach to positioning where items don't need to align vertically. From what I can tell, Bootstrap, Bulma, Lightning Design use one or the other (flexbox). Setting explicit widths on flexbox items to accomplish the same layout grid provides automatically feels like a hack to me, so I'd love to see support for both flex and grid.

- Updated the templates to ignore empty space after class attribute. 
- Removed all modifiers, except for `pf-m-align-right`. Let me know what you think.
- Updated documentation to reflect those changes
- Rebased w/master

michael-coker commented:
@andresgalante I'm a fan of utility classes. I have a handful in my application for common generic patterns in the app that aren't part of PF, so that I'm not repeating CSS, and in a lot of cases, it keeps me from having to write CSS at all.

I like your concept of building layouts based on what they are, and not how they're built for the reason that you gave - if there is a better way to achieve the layout in the future, we can change the layout without modifying the API. One option could be to create mixins for utility classes (or just mixins that aren't attached to a utility class, if we don't want utility classes), and apply the mixin to the utility class itself, as well as layout specific classes. So like, `@mixin pf-l-flex { display: flex; }`, `.pf-l-flex { @include pf-l-flex; }`, and `.pf-l-toolbar { @include pf-l-flex; }`. That would allow us to keep the layout specific classes without having to repeat code.

matthewcarleton commented:
I like what @michael-coker is proposing here for the layouts. That makes a lot of sense. Generally we want to avoid mixins but I don't think that rule needs to apply here.

andresgalante commented:
@michael-coker Your idea makes sense, by using a mixing we'll end up with a single source of truth by repetition of code. It's a cleaner API.

But maybe having just one grid and one flex layout will simplify things further. We will end up with things like gallery on the side but eliminate the need for the toolbar, split and level.

@mattnolting @michael-coker @matthewcarleton  let's talk about it today and come up with a decision.
